internal/services/authorization/role_definition_resource.go

```diff
@@ -9,365 +9,416 @@ import (
 	"github.com/hashicorp/go-uuid"
-	"github.com/hashicorp/terraform-provider-azurerm/helpers/tf"
-	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
-	"github.com/hashicorp/terraform-provider-azurerm/internal/services/authorization/azuresdkhacks"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/services/authorization/migration"
-	"github.com/hashicorp/terraform-provider-azurerm/internal/services/authorization/parse"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
-	"github.com/hashicorp/terraform-provider-azurerm/internal/timeouts"
-	"github.com/hashicorp/terraform-provider-azurerm/utils"
 )
 
-func resourceArmRoleDefinition() *pluginsdk.Resource {
-	return &pluginsdk.Resource{
-		Create: resourceArmRoleDefinitionCreate,
-		Read:   resourceArmRoleDefinitionRead,
-		Update: resourceArmRoleDefinitionUpdate,
-		Delete: resourceArmRoleDefinitionDelete,
+type ArmRoleDefinitionResource struct{}
 
-		Importer: pluginsdk.ImporterValidatingResourceId(func(id string) error {
-			_, err := parse.RoleDefinitionId(id)
-			return err
-		}),
+var _ sdk.ResourceWithUpdate = ArmRoleDefinitionResource{}
+var _ sdk.ResourceWithStateMigration = ArmRoleDefinitionResource{}
 
-		SchemaVersion: 1,
-		StateUpgraders: pluginsdk.StateUpgrades(map[int]pluginsdk.StateUpgrade{
-			0: migration.RoleDefinitionV0ToV1{},
-		}),
+type RoleDefinitionModel struct {
+	RoleDefinitionId         string            `tfschema:"role_definition_id"`
+	Name                     string            `tfschema:"name"`
+	Scope                    string            `tfschema:"scope"`
+	Description              string            `tfschema:"description"`
+	Permissions              []PermissionModel `tfschema:"permissions"`
+	AssignableScopes         []string          `tfschema:"assignable_scopes"`
+	RoleDefinitionResourceId string            `tfschema:"role_definition_resource_id"`
+}
 
-		Timeouts: &pluginsdk.ResourceTimeout{
-			Create: pluginsdk.DefaultTimeout(30 * time.Minute),
-			Read:   pluginsdk.DefaultTimeout(5 * time.Minute),
-			Update: pluginsdk.DefaultTimeout(60 * time.Minute),
-			Delete: pluginsdk.DefaultTimeout(30 * time.Minute),
-		},
+type PermissionModel struct {
+	Actions        []string `tfschema:"actions"`
+	NotActions     []string `tfschema:"not_actions"`
+	DataActions    []string `tfschema:"data_actions"`
+	NotDataActions []string `tfschema:"not_data_actions"`
+}
 
-		Schema: map[string]*pluginsdk.Schema{
-			"role_definition_id": {
-				Type:     pluginsdk.TypeString,
-				Optional: true,
-				Computed: true,
-				ForceNew: true,
-			},
+func (r ArmRoleDefinitionResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"role_definition_id": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+			Computed: true,
+			ForceNew: true,
+		},
```

Comment:
Can we add a validation, i.e. Any(UUID, EmptyString)?

Reply:
Can this be an empty string?

Reply:
If user specified it as empty string, it should work as same as not specified, the provider will generate a new one for it.